feat/integration-cloudflare-workers (#182)
xataio/mdx-docs: 070-Integrations/024-cloudflare-workers.mdx

---
title: Xata with Cloudflare Workers
navTitle: Cloudflare Workers
keywords: ['cloudflare', 'integration']
description: Xata with Cloudflare Workers
slug: integrations/cloudflare
published: true
---

[Cloudflare](https://www.cloudflare.com/en-gb/) is a content delivery and security service that optimizes website performance and mitigates security risks for sites. [Cloudflare Workers](https://workers.cloudflare.com/) allow you to run code directly on Cloudflare's global network, reducing latency and improving performance. They're ideal for tasks like customizing web content, handling API requests, and modifying web traffic, all executed closer to the user for faster response times.

> You'll need a Cloudflare account to deploy your application.

> The completed [Cloudflare and Xata code](https://github.com/xataio/examples/tree/main/apps/getting-started-cloudflare-workers) is available in the [Xata `examples` repo](https://github.com/xataio/examples) on GitHub.

Author: Where should the code for this live in the examples repo? `getting-started-cloudflare-workers` probably doesn't make since now since it's an integration

Author: The code is here: https://github.com/xataio/examples/pull/554

It only has the cloudflare dependencies, it assumes the user will run `xata init`

### Before you begin

- [Install the Xata CLI and authenticate](https://xata.io/docs/getting-started/installation)

### Create a new Cloudflare app

> Choose the hello world example and accept the default prompt options.

Author: ```suggestion
Choose the "Hello World" example and accept the default prompt options.
```

```sh
npm create cloudflare@latest xata-cloudflare
```

After the command has completed, navigate to the `xata-cloudflare` directory, and run the application locally:

```sh
npm run wrangler dev
```

By default, the application runs on `http://localhost:8787`. When visiting this URL "Hello World" should be shown in your browser.

### Deploy your app remotely

```sh
npx wrangler deploy
```

Preview your Worker at `<YOUR_WORKER>.<YOUR_SUBDOMAIN>.workers.dev.` When you visit this URL in your browser you should see "Hello World" returned.

### Install Xata Client

Inside your project root, run the following:

```sh
xata init
```

Accept the defaults. By the end you should have a `.env` file generated for you within your app folder. This file will contain the correct credentials to access your database from your Cloudflare worker.

```bash title=".env"
XATA_API_KEY=YOUR_API_KEY_HERE
XATA_BRANCH=main
```

### Initialize Xata Client

There are two ways of initiailizing Xata client in Cloudflare workers:

1.  ES Modules Syntax (Recommended)
2.  Service Workers Syntax (Deprecated)

> #### ES Modules Syntax (Recommended)

Author: Should these be numbered or fine as is?

    return new Response(`Total Posts: ${posts.length}`);
  }
};
```

#### Service Workers Syntax (Deprecated)

```ts title="src/index.ts"
import { XataClient } from './xata';

async function handler(request: Request) {
  const xata = new XataClient({
    apiKey: XATA_API_KEY,
    branch: XATA_BRANCH
  });
  // Note that the table name "Posts" may vary
  // depending on the shape of your schema
  const posts = await xata.db.Posts.getAll();
  return new Response(`Total Posts: ${posts.length}`);
}

// Initialize Worker
addEventListener('fetch', (event) => {
  event.respondWith(handler(event.request));
});
```

### Local environment variables

Copy your Xata environment values into a new file called .dev.vars at the root of your project. This allows Cloudflare access to them in dev mode.

```bash title=".dev.vars"
XATA_API_KEY=YOUR_API_KEY_HERE
XATA_BRANCH=main
```

### Remote environment variables

In order to make these environment variables accessible in your worker when it is deployed remotely, there are two options:

1.  One click guided workflow from the Cloudflare dashboard (Recommended)
2.  Manually through the Cloudflare dashboard

> #### One click guided workflow (Recommended)

Author: Should these be numbered or fine as is?

### Before you begin

- Ensure you [install the Xata CLI and authenticate](https://xata.io/docs/getting-started/installation)

### Create a new Cloudflare app

Choose the "Hello World" example and accept the default prompt options.

```sh
npm create cloudflare@latest getting-started-cloudflare-workers
```

After the command has completed, navigate to the `getting-started-cloudflare-workers` directory, and run the application locally:

```sh
npm run wrangler dev
```

By default, the application runs on `http://localhost:8787`. When visiting this URL "Hello World" should be shown in your browser.

### Deploy your app remotely

```sh
npx wrangler deploy
```

Preview your Worker at `<YOUR_WORKER>.<YOUR_SUBDOMAIN>.workers.dev.` When you visit this URL in your browser you should see "Hello World" returned.

### Install Xata Client

Inside your project root, run the following:

```sh
xata init
```

Accept the defaults. By the end you should have a `.env` file generated for you within your app folder. This file will contain the correct credentials to access your database from your Cloudflare worker.

```bash title=".env"

> XATA_API_KEY=YOUR_API_KEY_HERE

Reviewer: Noob question: How do users know where to go to get this key?

Reviewer: Ahh we've expected the user to go through the "before you begin" bit [here](https://xata-docs-pr-182.vercel.app/docs/integrations/cloudflare#before-you-begin), and at the bottom describe how to get the API key. 🤔

XATA_BRANCH=main
```

### Initialize Xata Client

> There are two ways of initiailizing Xata client in Cloudflare workers:

Author: I think most people would choose the recommended method, but there may be some user's who are still using Service Workers despite that being deprecated, and I guess we want to accommodate those users.

}

// Initialize Worker
addEventListener('fetch', (event) => {
  event.respondWith(handler(event.request));
});
```

### Local environment variables

Copy your Xata environment values into a new file called .dev.vars at the root of your project. This allows Cloudflare access to them in dev mode.

```bash title=".dev.vars"
XATA_API_KEY=YOUR_API_KEY_HERE
XATA_BRANCH=main
```

### Remote environment variables

In order to make these environment variables accessible in your worker when it is deployed remotely, there are two options:

1.  One click guided workflow from the Cloudflare dashboard (Recommended)
2.  Manually through the Cloudflare dashboard

#### One click guided workflow (Recommended)

Going through this flow automatically sets the environment variables necessary to interact with your database from a remote Cloudflare worker.

1.  Navigate to the Cloudflare dashboard
2.  Navigate to your corresponding Worker
3.  Click Integrations
4.  Select Xata

![Adding environment variables with our Xata Cloudflare integration](images/cloudflare-integration-env.png)

Your endpoint response should also be available remotely at `<YOUR_WORKER>.<YOUR_SUBDOMAIN>.workers.dev.`

> #### Manually adding environment variables

Reviewer: Cheeky suggestion: 
We could choose not to document this manual flow directly here, and rather link out to Cloudflare's [own docs for this process](https://developers.cloudflare.com/workers/configuration/environment-variables/#add-environment-variables-via-the-dashboard).

This lets us:
1. encourage people to focus on the recommended path
2. reduce the overhead of updating text/screenshots whenever Cloudflare update their docs

---
title: Cloudflare Workers with Xata
navTitle: Cloudflare Workers
keywords: ['cloudflare', 'integration']
description: Use Cloudflare Workers with Xata to improve performance
slug: integrations/cloudflare
published: true
---

[Cloudflare](https://www.cloudflare.com/en-gb/) is a content delivery network that optimizes website performance and mitigates security risks for sites. [Cloudflare Workers](https://workers.cloudflare.com) enable you to run code directly on Cloudflare's network to reduce latency and improve performance, and can be used for tasks like customizing web content, handling API requests, or modifying web traffic.

> A complete [Cloudflare and Xata starter](https://github.com/xataio/examples/tree/main/apps/getting-started-cloudflare-workers) is available in the [Xata `examples` repo](https://github.com/xataio/examples) on GitHub.

Reviewer 2: ```suggestion
A complete [Cloudflare and Xata starter](https://github.com/xataio/examples/tree/main/apps/getting-started-cloudflare-workers) is available in the Xata `examples` repo on GitHub.
```

> The completed [Cloudflare and Xata code](https://github.com/xataio/examples/tree/main/apps/getting-started-cloudflare-workers) is available in the [Xata `examples` repo](https://github.com/xataio/examples) on GitHub.

Reviewer 2: ```suggestion
A complete [Cloudflare and Xata starter](https://github.com/xataio/examples/tree/main/apps/getting-started-cloudflare-workers) is available in the [Xata `examples` repo](https://github.com/xataio/examples) on GitHub.
```

### Before you begin

- Ensure you [install the Xata CLI and authenticate](https://xata.io/docs/getting-started/installation)

- Ensure you have a [basic project with Cloudflare workers running](https://xata.io/docs/integrations/cloudflare#create-a-new-cloudflare-app-optional)

### Install the Xata client

Run the following command in the root directory of your project:

```sh
xata init
```

Accept the default settings during the configuration process. After completion, a `.env` file will be generated in your app's folder. This file includes the necessary credentials for your Cloudflare Worker to access your database. It should look like this:

```bash title=".env"
XATA_API_KEY=<YOUR_API_KEY_HERE>
XATA_BRANCH=<YOUR_BRANCH_HERE>
```

A file called `.xatarc` should also be generated at the project root, with the following contents:

```base title=".xatarc"
{
  "databaseURL": <YOUR_DATABASE_URL_HERE>,
  "codegen": {
    "output": "src/xata.ts"
  }
}
```

> > **Warning**: You'll need to move these vars into a file called `.dev.vars` at your project root for local development. This file enables Cloudflare to access these values when running in development mode. For more information, see their [documentation](https://developers.cloudflare.com/workers/configuration/environment-variables/#interact-with-environment-variables-locally).

Reviewer 2: ```suggestion
> **Warning**: You'll need to move these vars into a file called `.dev.vars` at your project root for local development. This file enables Cloudflare to access these values when running in development mode. For more information, see [Cloudflare's documentation](https://developers.cloudflare.com/workers/configuration/environment-variables/#interact-with-environment-variables-locally).
```

});
```

### Environment variables

Environment variables keep API keys and sensitive data safe, allowing you to easily manage and change app settings.

To add environment variables to a remotely deployed Worker, you have two options:

- Use the one-click guided workflow provided in the Cloudflare dashboard (recommended).
- Alternatively, you can set the variables manually via the Cloudflare dashboard.

#### One click guided workflow (recommended)

To integrate a Cloudflare Worker with an existing Xata database, follow these steps in the Cloudflare dashboard:

1.  Navigate to the [Cloudflare dashboard](https://dash.cloudflare.com/).
2.  Click **Workers & Pages** > **Overview**
3.  Choose a Worker from your Cloudflare account to integrate with.
4.  Click **Integrations**.
5.  Select the Xata card and click **Add Integration**. You're redirected to the "Add Xata to Worker" flow.
    ![Guided Xata Cloudflare integration](images/cloudflare-integration-env.png)
6.  Follow the flow to grant permissions, authorize Cloudflare, and configure your secrets. Choose a Xata workspace to connect with, and select the desired database and branch.
7.  Click **Add Integration**.

> The secrets are added to the environment variables of the Worker. Your endpoint response should be available remotely at `<YOUR_WORKER>.<YOUR_SUBDOMAIN>.workers.dev`

Reviewer 2: ```suggestion
The secrets are added to the environment variables of the Worker. Your endpoint response should be available remotely at `<YOUR_WORKER>.<YOUR_SUBDOMAIN>.workers.dev`.
```

---
title: Get started with Cloudflare and Xata
navTitle: Cloudflare Workers
keywords: ['cloudflare', 'full-text search']
description: Use Cloudflare and Xata
slug: getting-started/cloudflare
published: true
---

You will need a cloudflare account to deploy your application.

The completed [Cloudflare and Xata code](https://github.com/xataio/examples/tree/main/apps/getting-started-cloudflare-workers) for this
guide is available via the [Xata `examples` repo](https://github.com/xataio/examples) on GitHub.

## Before you begin

> Install the Xata CLI:

Author: Maybe we can just offload this whole section to a link to https://xata-docs-pr-182.vercel.app/docs/getting-started/installation

```sh
npm install -g @xata.io/cli
```

Authenticate the Xata CLI with your Xata account. If you don't already have an account, you can use the
same workflow to sign up for a new account. Run the following command to begin the authentication workflow:

```sh
xata auth login
```

On completion, the command will create a new API key for your user account, which you should see in the \[account
settings page within the Xata UI]\[1]. That key will also be stored locally on your computer (the location might vary for
each OS). It looks like this:

```toml
# .config/xata/credentials
[default]
apiKey=YOUR_API_KEY_HERE
```

> ## Create a new Cloudflare app

Author: I think this section can stay

1.  Create a new Cloudflare workers endpoint, accepting the default prompt options:

    ```sh
    npm create cloudflare@latest
    ```

2.  Once the command has completed, go to the `getting-started-cloudflare-workers` directory, and run the application:

    ```sh
    $ npm run wrangler dev
    ```

By default, the application will run on `http://localhost:8787`. When visiting this URL "Hello World" should be shown in your browser.

> ## Create a new database

Author: This section should probably go, would be great to offload it to a link. I think the CSV import stuff should definitely go
